File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/graphql/schema/GQLTypeValidationRun.java
package net.nemerosa.ontrack.boot.graphql.schema;

import graphql.schema.GraphQLNonNull;
import graphql.schema.GraphQLObjectType;
import graphql.schema.GraphQLTypeReference;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.structure.PropertyService;
import net.nemerosa.ontrack.model.structure.Signature;
import net.nemerosa.ontrack.model.structure.ValidationRun;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

import static graphql.Scalars.GraphQLInt;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static graphql.schema.GraphQLObjectType.newObject;
import static net.nemerosa.ontrack.boot.graphql.support.GraphqlUtils.stdList;

@Component
public class GQLTypeValidationRun extends AbstractGQLProjectEntity<ValidationRun> {

    public static final String VALIDATION_RUN = "ValidationRun";

    private final GQLTypeValidationRunStatus validationRunStatus;

    public GQLTypeValidationRun(GQLTypeValidationRunStatus validationRunStatus,
                                PropertyService propertyService,
                                GQLTypeProperty property,
                                List<GQLProjectEntityFieldContributor> projectEntityFieldContributors) {
        super(ValidationRun.class, ProjectEntityType.VALIDATION_RUN, propertyService, property, projectEntityFieldContributors);
        this.validationRunStatus = validationRunStatus;
    }

    @Override
    public GraphQLObjectType getType() {
        return newObject()
                .name(VALIDATION_RUN)
                .withInterface(projectEntityInterface())
                .fields(projectEntityInterfaceFields())
                // Build
                .field(
                        newFieldDefinition()
                                .name("build")
                                .description("Associated build")
                                .type(new GraphQLNonNull(new GraphQLTypeReference(GQLTypeBuild.BUILD)))
                                .build()
                )
                // Promotion level
                .field(
                        newFieldDefinition()
                                .name("validationStamp")
                                .description("Associated validation stamp")
                                .type(new GraphQLNonNull(new GraphQLTypeReference(GQLTypeValidationStamp.VALIDATION_STAMP)))
                                .build()
                )
                // Run order
                .field(
                        newFieldDefinition()
                                .name("runOrder")
                                .description("Run order")
                                .type(GraphQLInt)
                                .build()
                )
                // Validation statuses
                .field(
                        newFieldDefinition()
                                .name("validationRunStatuses")
                                .description("List of validation statuses")
                                .type(stdList(validationRunStatus.getType()))
                                .build()
                )
                // OK
                .build();

    }

    @Override
    protected Optional<Signature> getSignature(ValidationRun entity) {
        return Optional.ofNullable(entity.getLastStatus().getSignature());
    }
}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLTypeValidationRun.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars.GraphQLInt
import graphql.schema.GraphQLNonNull
import graphql.schema.GraphQLObjectType
import graphql.schema.GraphQLObjectType.newObject
import graphql.schema.GraphQLTypeReference
import net.nemerosa.ontrack.extension.api.ExtensionManager
import net.nemerosa.ontrack.graphql.schema.authorizations.GQLInterfaceAuthorizableService
import net.nemerosa.ontrack.graphql.support.booleanArgument
import net.nemerosa.ontrack.graphql.support.listType
import net.nemerosa.ontrack.graphql.support.toNotNull
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.structure.RunInfoService
import net.nemerosa.ontrack.model.structure.Signature
import net.nemerosa.ontrack.model.structure.ValidationRun
import net.nemerosa.ontrack.model.support.FreeTextAnnotatorContributor
import org.springframework.stereotype.Component

@Component
class GQLTypeValidationRun(
    creation: GQLTypeCreation,
    private val validationRunStatus: GQLTypeValidationRunStatus,
    projectEntityFieldContributors: List<GQLProjectEntityFieldContributor>,
    private val runInfo: GQLTypeRunInfo,
    private val runInfoService: RunInfoService,
    private val validationRunData: GQLTypeValidationRunData,
    private val projectEntityInterface: GQLProjectEntityInterface,
    private val gqlInterfaceAuthorizableService: GQLInterfaceAuthorizableService,
    freeTextAnnotatorContributors: List<FreeTextAnnotatorContributor>,
    extensionManager: ExtensionManager,
) : AbstractGQLProjectEntity<ValidationRun>(
    ValidationRun::class.java,
    ProjectEntityType.VALIDATION_RUN,
    projectEntityFieldContributors,
    creation,
    freeTextAnnotatorContributors,
    extensionManager,
) {

    override fun getTypeName() = VALIDATION_RUN

    override fun createType(cache: GQLTypeCache): GraphQLObjectType =
        newObject()
            .name(VALIDATION_RUN)
            .withInterface(projectEntityInterface.typeRef)
            .fields(projectEntityInterfaceFields())
            // Build
            .field {
                it
                    .name("build")
                    .description("Associated build")
                    .type(GraphQLNonNull(GraphQLTypeReference(GQLTypeBuild.BUILD)))
            }
            // Promotion level
            .field {
                it
                    .name("validationStamp")
                    .description("Associated validation stamp")
                    .type(GraphQLNonNull(GraphQLTypeReference(GQLTypeValidationStamp.VALIDATION_STAMP)))
            }
            // Run order
            .field {
                it
                    .name("runOrder")
                    .description("Run order")
                    .type(GraphQLInt)
            }
            // Validation statuses
            .field {
                it
                    .name("validationRunStatuses")
                    .description("List of validation statuses")
                    .type(listType(validationRunStatus.typeRef))
                    .argument(
                        booleanArgument(
                            "lastOnly",
                            "True to get only the last status only."
                        )
                    )
                    .dataFetcher { environment ->
                        val validationRun = environment.getSource<ValidationRun>()
                        val lastOnly = environment.getArgument<Boolean>("lastOnly") ?: false
                        val statuses = if (lastOnly) {
                            validationRun.validationRunStatuses.take(1)
                        } else {
                            validationRun.validationRunStatuses
                        }
                        statuses.map { status ->
                            GQLTypeValidationRunStatus.Data(
                                validationRun,
                                status
                            )
                        }
                    }
            }
            // Last status
            .field {
                    it.name("lastStatus")
                        .description("Last validation status for this run")
                        .type(validationRunStatus.typeRef.toNotNull())
                        .dataFetcher { env ->
                            val validationRun: ValidationRun = env.getSource()
                            GQLTypeValidationRunStatus.Data(
                                validationRun,
                                validationRun.lastStatus
                            )
                        }
            }
            // Run info
            .field {
                it.name("runInfo")
                    .description("Run info associated with this validation run")
                    .type(runInfo.typeRef)
                    .runInfoFetcher<ValidationRun> { runInfoService.getRunInfo(it) }
            }
            // Data
            .field {
                it.name("data")
                    .description("Data associated with the validation run")
                    .type(validationRunData.typeRef)
            }
            // Authorizations
            .apply {
                gqlInterfaceAuthorizableService.apply(this, ValidationRun::class)
            }
            // OK
            .build()

    override fun getSignature(entity: ValidationRun): Signature? {
        return entity.lastStatus.signature
    }

    companion object {
        const val VALIDATION_RUN = "ValidationRun"
    }
}
